File: code/08-graphs/python/dag_shortest_paths.py
"""Same DAGShortestPaths as DagShortestPaths.java/dag_shortest_paths.c --
newly written in Python, no canonical source exists."""
from graph import Graph
from shortest_paths import INF, reconstruct_path
from topological_sort import kahn


# snippet:dag-shortest-paths:start
def dag_shortest_paths(g, s):
    order = kahn(g)
    if order is None:
        raise ValueError("not a DAG")
    n = g.vertices()
    dist = [INF] * n
    parent = [-1] * n
    dist[s] = 0
    for u in order:
        if dist[u] == INF:
            continue
        for v, w in g.edges_from(u):
            cand = dist[u] + w
            if cand < dist[v]:
                dist[v] = cand
                parent[v] = u
    return dist, parent
# snippet:dag-shortest-paths:end


if __name__ == "__main__":
    name = "sabc"
    g = Graph(4, True)
    for u, v, w in [(0,1,3),(0,2,2),(1,3,-4),(2,3,1)]:
        g.add_edge(u, v, w)
    dist, parent = dag_shortest_paths(g, 0)
    print("dist:", [int(x) for x in dist])
    path = reconstruct_path(parent, dist, 3)
    print("path to c:", " -> ".join(name[v] for v in path))

File: code/08-graphs/java/DagShortestPaths.java
import java.util.*;

// New: no canonical source (lecture-notes/code/lecture08/java/ShortestPaths.java
// only has Dijkstra/Bellman-Ford). Reuses TopologicalSort.kahn verbatim and
// adds the Relax loop the DAGShortestPaths pseudocode describes.
public final class DagShortestPaths {
    public record Result(long[] dist, int[] parent) {}
    private DagShortestPaths() {}

    // snippet:dag-shortest-paths:start
    public static Result dagShortestPaths(Graph g, int s) {
        int[] order = TopologicalSort.kahn(g).orElseThrow(() -> new IllegalArgumentException("not a DAG"));
        int n = g.vertices();
        long[] dist = new long[n]; int[] parent = new int[n];
        Arrays.fill(dist, ShortestPaths.INF); Arrays.fill(parent, -1); dist[s] = 0;
        for (int u : order) {
            if (dist[u] == ShortestPaths.INF) continue;
            for (Graph.Edge e : g.edgesFrom(u)) {
                long cand = dist[u] + e.weight();
                if (cand < dist[e.to()]) { dist[e.to()] = cand; parent[e.to()] = u; }
            }
        }
        return new Result(dist, parent);
    }
    // snippet:dag-shortest-paths:end

    public static void main(String[] args) {
        String[] name = {"s", "a", "b", "c"};
        Graph g = new Graph(4, true);
        long[][] e = {{0,1,3},{0,2,2},{1,3,-4},{2,3,1}};
        for (long[] x : e) g.addEdge((int)x[0], (int)x[1], x[2]);
        Result r = dagShortestPaths(g, 0);
        System.out.println("dist: " + Arrays.toString(r.dist()));
        int[] path = ShortestPaths.path(new ShortestPaths.Result(r.dist(), r.parent(), false), 3);
        StringBuilder sb = new StringBuilder();
        for (int v : path) { if (sb.length() > 0) sb.append(" -> "); sb.append(name[v]); }
        System.out.println("path to c: " + sb);
    }
}
